common/migration/default-comparison.go

```diff
@@ -0,0 +1,157 @@
+			return false
+		}
+		// else they actually match in both having the same error
+		scope.IncCounter(metrics.ShadowerComparisonMatchCount)
+		return true
+	}
+
+	if backgroundErr != nil && activeErr == nil {
+		log.Warn("mismatched result with background flow returning an error, whereas the active flow succeeded",
+			tag.Error(backgroundErr))
+		scope.IncCounter(metrics.ShadowerComparisonDifferenceCount)
+		return false
+	}
+
+	if activeErr != nil && backgroundErr == nil {
+		log.Warn("mismatched result with active flow returning an error, whereas the background flow succeeded",
+			tag.Error(activeErr))
+		scope.IncCounter(metrics.ShadowerComparisonDifferenceCount)
+		return false
+	}
+
+	// so, the problem here is that cmp.Diff doesn't love to receive pointers, and tends to panic
+	// for a fairly wide range of inputs when it hits a panic. This is not a test system, but rather
+	// an online flow, and so the use of a generic reflection tool is both slow and not particularly suited
+	// to this usecase, and quite dangerous insofar as that it panics a lot.
+	//
+	// That said, the ability to spot mapping differences, serialization differences and similar through its deep
+	// reflection is invaluable, so while this is a dangerous tradeoff, it's chosen here carefully, with the knowledge
+	// that this comparison is to be used in a context where panics are recovered and precautions such as the below are
+	// taken to try and catch the egregious cases which will cause panics.
+
+	// Firstly, use a bit of reflection to work out of the given types are pointers
+	// and try and derefrence them.
+	backgroundVal, d1, err := followPtrsAndReturnValue(backgroundRes)
+	if err != nil {
+		return false // fail to compare due to being unable to deref safely
+	}
+
+	activeVal, d2, err := followPtrsAndReturnValue(activeRes)
+	if err != nil {
+		return false // fail to compare due to being unable to deref safely
+	}
+
+	if d1 != d2 {
+		log.Warn("shadow flow got mismatched results, the results have a different number of pointers. Probably one of the values is a pointer and the other a value")
+		scope.IncCounter(metrics.ShadowerComparisonDifferenceCount)
+		return false
+	}
+
+	if backgroundVal == nil && activeVal == nil {
+		scope.IncCounter(metrics.ShadowerComparisonMatchCount)
+		return true
+	}
+
+	// and then go through and short-circuit if anything doesn't appear to be a struct
+	backgroundType := reflect.TypeOf(backgroundVal)
+	activeType := reflect.TypeOf(activeVal)
+
+	if backgroundType == nil || backgroundType.Kind() != reflect.Struct ||
+		activeType == nil || activeType.Kind() != reflect.Struct {
+
+		log.Warn("shadow flow got mismatched results, the types to compare are not struct or pointer to structs")
+		scope.IncCounter(metrics.ShadowerComparisonDifferenceCount)
+		return false
+	}
+
+	diff := cmp.Diff(
+		backgroundVal,
+		activeVal,
+		cmpopts.IgnoreUnexported(backgroundVal),
+		cmpopts.IgnoreUnexported(activeVal))
```

Comment:
hmmm: https://pkg.go.dev/github.com/google/go-cmp@v0.6.0/cmp/cmpopts#IgnoreUnexported
> ... only ignores the immediate unexported fields of a struct, ...

that "immediate" makes me worry that this only ignores at the top level, not recursively.  and the implementation doesn't reduce that worry, at a glance.

if this isn't recursive I think it might be (very) problematic, and we might need to do something else.

Reply:
yeah, good spotting, I didn't notice that. Let me add a test and determine if that's true, if it is, 100% agree and I'll just drop cmp.Diff entirely and maybe just use a JSON differ or something

Reply:
you're correct, ty, good spotting